c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -161,10 +162,62 @@ impl<'tcx> TypeChecker<'tcx, '_> {
+    pub fn visit_operand(&mut self, op: &Operand<'tcx>) -> LTy<'tcx> {
         match *op {
             Operand::Copy(pl) | Operand::Move(pl) => self.visit_place(pl),
             Operand::Constant(ref c) => {
+                if c.ty().is_any_ptr() {
+                    if let Some(alloc_id) = const_alloc_id(c) {
+                        if let Some(did) = find_static_for_alloc(&self.acx.tcx(), alloc_id) {
+                            let lty = assign_origins(
+                                self.ltcx,
+                                self.hypothesis,
+                                self.facts,
+                                self.maps,
+                                &self.acx.gacx.adt_metadata,
+                                &self.acx.gacx.static_tys[&did],
+                            );
+
+                            for l in lty.iter() {
+                                let static_origin = self.static_origin;
+                                let mut add_subset_base = |pl: Origin, rv: Origin| {
+                                    let point = self.current_point(SubPoint::Mid);
+                                    self.facts.subset_base.push((rv, pl, point));
+                                };
+                                if let Some(origin) = l.label.origin {
+                                    // constrain this origin to be 'static
+                                    eprintln!("constraining origin {origin:?} to 'static lifetime");
+                                    add_subset_base(static_origin, origin);
+                                    add_subset_base(origin, static_origin)
+                                }
+                                for (op, origin) in l.label.origin_params {
+                                    // constrain this origin to be 'static
+                                    eprintln!("constraining origin {op:?} ({origin:?}) to 'static lifetime");
+                                    add_subset_base(static_origin, *origin);
+                                    add_subset_base(*origin, static_origin);
+                                }
+                            }
+
+                            eprintln!("NEW STATIC LTY: {:?}", c.ty().kind());
+                            let pointer_id = self.acx.type_of(op).label;
+                            let perm = self.hypothesis[pointer_id];
+                            let args = self.ltcx.mk_slice(&[lty]);
+                            assert!(matches!(
+                                *c.ty().kind(),
+                                TyKind::Ref(..) | TyKind::RawPtr(..)
+                            ));
+
+                            // Polonius does not appear to issue loans for statics
+                            let label = Label {
+                                origin: None,
+                                origin_params: &[],
+                                perm,
+                            };
```

Comment:
Rustc doesn't issue loans for these, but it does assign an origin (at least in cases where the `Operand::Constant` for the static has ref type, which we generalize to both ref and raw pointer types).  Rustc also doesn't set any constraints based on the operand or rvalue directly, but the origin produced here can be involved in other constraints, such as the one produced by the enclosing `StatementKind::Assign`.

Reply:
So we should generate a fresh origin here as well to assign to the pointer type that `c` in `Operand::Constant(ref c)` is?

Reply:
Added in 6ccfd803